### PR TITLE
Moved if statement to assign task completion before saving task to user

### LINKED ISSUE #4891: Todo task created via user/tasks/{id}/up is not marked as completed
At the moment, when creating a new todo task via the `user/tasks/{id}/up` route, the user gets xp and gold, but the todo task remains incomplete. Should the default not be that the todo task also gets marked as complete?

### PR DESCRIPTION
**Pull request by TheHollidayInn**
It seems like the if statement need to be moved up to fix https://github.com/HabitRPG/habitrpg/issues/4891 .

Here are some tests:

**New Todo with Up**
screen shot 2015-04-07 at 7 31 58 pm
screen shot 2015-04-07 at 7 31 36 pm

**New Todo with Down**
screen shot 2015-04-07 at 7 32 27 pm
screen shot 2015-04-07 at 7 32 21 pm

**New Habit Down**
screen shot 2015-04-07 at 7 33 48 pm

**New Habit Up (Notice the previous habit is under the new green habit)**
screen shot 2015-04-07 at 7 34 11 pm

### CODE REVIEW

**Comment by crookedneighbor** [comment]
@kusold you might be interested in this.

This change will allow the task to be scored on completion, rather than requiring two separate calls (one to mark as complete and one to score), like we had to do for the todoist integration.

**Comment by Alys** [comment]
It's just occurred to me, before this is merged, we should make some attempt to publicise that this is going to be done. It's not polite to our third-party developers to change the API without warning, even when that change is beneficial. They might have work-arounds for the lack of this feature that would cause problems when this feature is introduced.

See also [API changes aren't visible enough to third-party developers/partners - Issue #4432](https://github.com/HabitRPG/habitrpg/issues/4432). I suggest that anyone with ideas about how to best publicise API changes should drop in there and add comments.

**Comment by crookedneighbor** [comment]
@Alys That is true. Until the todoist-habitrpg module is updated, this would make it so my todos in Todoist are double scored, once for being completed, and once using the score call.

Unfortunately, I have no ideas for publicizing it. :frowning:

**Comment by crookedneighbor** [comment]
Also, I want tests around this behavior.

**Comment by Alys** [comment]
I've labelled this as "suggestion-discussion" until we sort out how to notify third-party developers about the change to the API (or release a new version of the API). https://github.com/HabitRPG/habitrpg/issues/4432

**Comment by crookedneighbor** [comment]
@TheHollidayInn I'm still interested in having this merged in, because I believe it is the correct behavior. But it still needs tests.

**Comment by TheHollidayInn** [comment]
Can do. This was before I knew how to make api tests.

I rebased my branch, and now, I am getting this error: 

```
/Users/krh121791/habitrpg/node_modules/mixpanel/lib/mixpanel-node.js:19
        throw new Error("The Mixpanel Client needs a Mixpanel token: `init(tok
              ^
Error: The Mixpanel Client needs a Mixpanel token: `init(token)`
```

Do you have any idea what this is?

**Comment by Alys** [comment]
Update your local config.json with the latest additions to the example config.json. That should fix it. 

Note also that package.json has had some recent additions so you might need to do `npm install`

**Comment by TheHollidayInn** [comment]
Thanks, @Alys !

How do these tests look?

**Comment by crookedneighbor** [comment]
Good so far. It should also test that it marks an existing uncompleted todo as completed when scored up and a previously completed todo is marked as incomplete when scored down.

**Comment by TheHollidayInn** [comment]
Here you go.

File: test/api/todos.coffee [hunk 1]
```diff
@@ -184,3 +251,34 @@ describe "Todos", ->
           expect(body.err).to.equal "Task not found."
           done()
 
+describe "habits", ->
+
+  it "Creates a completed a habit when using up url", (done) ->
```
Comment (crookedneighbor) [inline test/api/todos.coffee]: Habits don't complete

File: test/api/todos.coffee [hunk 2]
```diff
@@ -184,3 +251,34 @@ describe "Todos", ->
...
+        description: "testDesc"
+    ).end (res) ->
+      expectCode res, 200
+      request.get(baseURL + "/user/tasks/withUp")
+      .send().end (res) ->
+        upHabit = res.body
+        expect(upHabit.completed).to.equal true
+        done()
+
+  it "Creates an uncompleted todo when using down url", (done) ->
```
Comment (crookedneighbor) [inline test/api/todos.coffee]: Again, habits don't have a complete/incomplete state

File: test/api/todos.coffee [hunk 3]
```diff
@@ -184,3 +251,34 @@ describe "Todos", ->
           expect(body.err).to.equal "Task not found."
           done()
 
+describe "habits", ->
+
+  it "Creates a completed a habit when using up url", (done) ->
+    upHabit = undefined
+    request.post(baseURL + "/user/tasks/withUp/up").send(
```
Comment (crookedneighbor) [inline test/api/todos.coffee]: I don't think you're actually creating a habit here. You're actually scoring the todo you created earlier in the test.

File: test/api/todos.coffee (hunk 2 above)
Comment (TheHollidayInn) [inline test/api/todos.coffee]: The complete/incomplete state was part of the issue request. So, I used test to mimic what the user was doing. They were confused between the creation of a habit vs todo when setting the complete status and using the up url.

Comment (TheHollidayInn) [inline test/api/todos.coffee]: Nevermind. I see what you mean. I am testing it now.

**Comment by TheHollidayInn** [comment]
Okay, I have updated the tests.

File: test/api/todos.coffee [hunk 4]
```diff
@@ -97,6 +97,36 @@ describe "Todos", ->
           expect(todo.value).to.equal 0
           done()
 
+      it "Creates a completed a todo when using up url", (done) ->
+        upTodo = undefined
+        request.post(baseURL + "/user/tasks/withUp/up").send(
+            type: "todo"
+            text: "testTitle"
+            complete: true
```
Comment (crookedneighbor) [inline test/api/todos.coffee]: I think you should not pass this parameter if you're checking that the scoring route does the work of completing or uncompleting it.

File: test/api/todos.coffee [hunk 5]
```diff
@@ -97,6 +97,36 @@ describe "Todos", ->
...
+            upTodo = res.body
+            expect(upTodo.completed).to.equal true
+            done()
+
+      it "Creates an uncompleted todo when using down url", (done) ->
+        downTodo = undefined
+        request.post(baseURL + "/user/tasks/withDown/down").send(
+            type: "todo"
+            text: "testTitle"
+            complete: true
```
Comment (crookedneighbor) [inline test/api/todos.coffee]: I feel like checking that the scoring route overrides what parameter the user passes in for complete should be a separate test.

Comment (TheHollidayInn) [inline test/api/todos.coffee]: So, I should have a create normal, a create to override, and the update test for todos below?

**Comment by TheHollidayInn** [comment]
How do these tests look?

**Comment by crookedneighbor** [comment]
Moved the tests to their own file: https://github.com/HabitRPG/habitrpg/commit/2acf7eb71e7cfa54d2e34f0a9738ffd140add40d

Notated toward your next tier.

**Pull request merged**